### PR TITLE
Re-enable `from_json` / `JsonToStructs`

### PR DESCRIPTION
**Pull request by andygrove**
Depends on https://github.com/rapidsai/cudf/pull/14309

This PR re-enables `from_json` / `JsonToStructs`. It was previously disabled (see https://github.com/NVIDIA/spark-rapids/issues/8558) because it can cause incorrect results in some cases, but we have similar issues when reading directly from JSON files (see xfailed cases in https://github.com/NVIDIA/spark-rapids/pull/9304).

Given that this expression does work correctly in many cases where the inputs are valid JSON, and that it is disabled by default, I would like to discuss adding this back in.

### CODE REVIEW

**Review by revans2 — COMMENTED** [review]
For me personally I would disable JSON reading and CSV reading for that matter by default, instead of compounding the problem.

**Comment by andygrove** [comment]
I am seeing two errors with the current code:

- Some valid JSON lines are being replaced with null when recoverWithNulls is enabled
- Also seeing the following exception in some cases

```
- E                   Caused by: java.lang.IllegalStateException: The input data didn't parse correctly and we read a different number of rows than was expected. Expected 512, but got 511
E                       at org.apache.spark.sql.rapids.GpuJsonToStructs.$anonfun$doColumnar$6(GpuJsonToStructs.scala:179)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 1]
```diff
@@ -3556,17 +3556,19 @@ object GpuOverrides extends Logging {
     expr[JsonToStructs](
       "Returns a struct value with the given `jsonStr` and `schema`",
       ExprChecks.projectOnly(
-        TypeSig.MAP.nested(TypeSig.STRING).withPsNote(TypeEnum.MAP,
-              "MAP only supports keys and values that are of STRING type"),
+        TypeSig.STRUCT.nested(TypeSig.all) +
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3559]: I don't think we really support all under a Struct. We don't support maps under a struct and we don't support UDTs under a struct. Could we be more explicit about what we do and do not support here?

**Review by revans2 — APPROVED** [review]

**Comment by andygrove** [comment]
build

**Pull request merged**